### ISSUE TITLE
Only one registry per view superclass

### ISSUE THREAD

**Issue by pascalbetz** [hanami/view#89]
According to comment https://github.com/lotus/view/blob/master/lib/lotus/view/rendering/registry.rb#L48 there should be only one registry per view superclass. 

According to my tests there is one registry per view. What is the reason for trying to have one registry per Superclass?

If no good reason i'd remove this code and make sure that there is one registry per view.

In addition: even if the code worked as described in the comments, subclass views would be overwritten by superclass views if they have the same format because they are keyed by format. See https://github.com/lotus/view/blob/master/lib/lotus/view/rendering/registry.rb#L104 and https://github.com/lotus/view/blob/master/lib/lotus/view/rendering/registry.rb#L115

**Comment by jodosha** [comment]
@pascalbetz The docs express the intended implementation. :wink: 
If you have spotted a different implementation, please open a PR with the fix.

The goal is to have **one registry per view superclass**.

```ruby
module Web::Views::Books
  # This should hold the registry
  class Show
    include Web::View
  end

  # This should NOT
  class JsonShow < Show
    format :json
  end
end
```

**Comment by jodosha** [comment]
@pascalbetz 

> In addition: even if the code worked as described in the comments, subclass views would be overwritten by superclass views if they have the same format because they are keyed by format. See

Let me explain the intent. That code is registering all the handled formats, according to the available templates.

There are a few cases to cover:

1. **One superclass, multiple templates.** We have just `Show`, but multiple templates for different formats like: `show.html.erb`, `show.json.erb`. So the view registers the `:html` and `:json` format at the boot time. If an CSV request comes in, we can't handle it, because `:csv` is a missing key in the registry.

2. **Multiple views, multiple templates.** We have `Show` that handles `show.html.erb` and `show.xml.erb`, than we have `JsonShow` that handles `show.json.erb`.

---

If we can improve how the registry is stored in memory, I'm totally open to this. As long the tests won't change at all. I spent a lot of time to make this to work and to discover these use cases.

Thank you! :wink:

**Comment by pascalbetz** [comment]
If you have 

```Show``` handling ```show.html.erb``` and ```ExtendedShow``` handling ```extendedshow.html.erb```  then the registry would override one or the other because the key is the format (```html``` in both cases).

So, independent of the fact that this might not be the intention of subclass views, i think the implementation and the idea of having one registry per superclass view is faulty.

Implementation: AFAIK currently there can be subclass views with same format which indicates that there is not just one registry per superclass
Idea: What is the exact idea behind the "one registry per superclass view" restriction? I think it adds complexity without benefit.

IMHO the changes in I think that https://github.com/lotus/view/pull/90 shows that there is multiple registries.

**Comment by jodosha** [comment]
@pascalbetz

> If you have
> 
> Show handling show.html.erb and ExtendedShow handling extendedshow.html.erb then the registry would override one or the other because the key is the format (html in both cases).

Now, I see.. Thanks for explaining :wink: 
Well, this use case isn't supported, and I don't see any reason why we should. Any idea?

> Idea: What is the exact idea behind the "one registry per superclass view" restriction? I think it adds complexity without benefit.

For a given action `Web::Controllers::Books::Show`, there is one corresponding `Web::Views::Books::Show` view. This last component is responsible to render the current asked format. One registry per superclass (aka `Show` view), helps to take the decision if to handle the rendering by itself, or delegate one of the specialized subclasses.

---

To recap:

  1. There is no need to support multiple templates for the same format (eg. `show.html.erb` and `extended_show.html.erb`. Multiple templates are supported for multiple MIME Types (aka formats). Eg. `show.html.erb` and `show.json.erb`.

  2. One registry per superclass is mandatory for the naming convention between an action and a view. See `Lotus::RenderingPolicy` in `lotusrb`.

---

@pascalbetz Does this helps to clarify the intent? That being said, if we have a more efficient way to store the registry in memory, it's more than welcome. :smile:

**Comment by pascalbetz** [comment]
Hello @jodosha 

I think my confusion stems from having subclasses for different formats vs. subclasses for different actions. If you are serving a request for a subclass view it will be loaded on request time when it is served from a different action. I think this is why you could not reproduce (sorry for the faulty description, in that case).


Now i find it confusing that the naming pattern breaks (ShowXml) when having different views for different formats. But i don't have a better idea right now. Except

```ruby
Foo::View::People::Index::Html
Foo::View::People::Index::Xml
Foo::View::People::Index::Json
```

Which would require a lot more typing and breaks as soon as the same view serves multiple formats.

> Now, I see.. Thanks for explaining :wink: 
> Well, this use case isn't supported, and I don't see any reason why we should. Any idea?

Perhaps a warning? Or documentation?

> For a given action Web::Controllers::Books::Show, there is one corresponding > Web::Views::Books::Show view. This last component is responsible to render the current asked format. 
> One registry per superclass (aka Show view), helps to take the decision if to handle the rendering by 
> itself, or delegate one of the specialized subclasses.

If there is already static information (the registry) why not have an "overall" registry?

**Comment by jodosha** [comment]
@pascalbetz 
> If you are serving a request for a subclass view it will be loaded on request time when it is served from a different action.

This is still unclear for me:

  * Where and how did you observed this lazy loading mechanism?
  * What does this means: "when it is served from a different action"? I assume that `Show` view is backed by `Show` action.

---

> If there is already static information (the registry) why not have an "overall" registry?

That could be a solution as well, but what are the gains? Lower memory footprint? Faster runtime lookups? More code flexibility?

:smile:

**Comment by jodosha** [comment]
@pascalbetz I'm closing this because not high priority.
